app/code/Magento/Inventory/Model/Stock/Command/GetProductQuantityInStock.php

```diff
@@ -0,0 +1,125 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Inventory\Model\Stock\Command;
+
+use Magento\Framework\App\ResourceConnection;
+use Magento\Inventory\Indexer\Alias;
+use Magento\Inventory\Indexer\IndexNameBuilder;
+use Magento\Inventory\Indexer\IndexNameResolverInterface;
+use Magento\Inventory\Indexer\StockItem\IndexStructure as StockItemIndex;
+use Magento\Inventory\Indexer\StockItemIndexerInterface;
+use Magento\Inventory\Setup\Operation\CreateReservationTable;
+use Magento\InventoryApi\Api\Data\ReservationInterface;
+use Magento\InventoryApi\Api\GetProductQuantityInStockInterface;
+
+/**
+ * Return Quantity of products available to be sold by Product SKU and Stock Id
+ *
+ * @see \Magento\InventoryApi\Api\GetProductQuantityInStockInterface
+ * @api
+ */
+class GetProductQuantityInStock implements GetProductQuantityInStockInterface
+{
+    /**
+     * @var ResourceConnection
+     */
+    private $resource;
+
+    /**
+     * @var IndexNameBuilder
+     */
+    private $indexNameBuilder;
+
+    /**
+     * @var IndexNameResolverInterface
+     */
+    private $indexNameResolver;
+
+    /**
+     * GetProductQuantityInStock constructor.
+     *
+     * @param ResourceConnection $resource
+     * @param IndexNameBuilder $indexNameBuilder
+     * @param IndexNameResolverInterface $indexNameResolver
+     */
+    public function __construct(
+        ResourceConnection $resource,
+        IndexNameBuilder $indexNameBuilder,
+        IndexNameResolverInterface $indexNameResolver
+    ) {
+        $this->resource = $resource;
+        $this->indexNameBuilder = $indexNameBuilder;
+        $this->indexNameResolver = $indexNameResolver;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $sku, int $stockId): float
+    {
+        $productQtyInStock = $this->getStockItemQty($sku, $stockId) - $this->getReservationQty($sku, $stockId);
+        return (float) $productQtyInStock;
+    }
```

Comment:
@aleron75  you raised a correct question. 
Whether MySQL logic should be placed in the Domain Service implementation.

we usually don't do that introducing intermediate level of abstraction.
For example, Repository does not make direct requests to MySQL table, what it does - it uses ResourceModels to get needed data.

Ideally, Implementations of Service Contracts should be *agnostic* to the Data Access Layer.

What I recommend to do in this particular case - I propose to introduce two Models and inject them into this Service.
- The first model would be responsible for retrieving StockItem Quantity (implementation of your getStockItemQty method).
- The second one would be responsible for getting Reservation Quantity (getReservationQty in current code)

So, you would end up in current service with next code:
```
$productQtyInStock = $this->stockItemQty->execute($sku, $stockId) - 
$this->reservationQty->execute($sku, $stockId);
return (float) $productQtyInStock;
```

Reply:
sounds good, just to be sure, by _"I propose to introduce two Models"_ you mean Resource Models, is it right?

Reply:
@aleron75  Correct, Resource models